### ISSUE TITLE
OneDrive can skip single file upload

### ISSUE THREAD

**Issue by B4dM4n** [ncw/rclone#3164]
#### What is the problem you are having with rclone?

When uploading a single file to a OneDrive folder, it can get skipped if the file already exists in the root folder and the destination folder does not exist.

This happens because [readMetaDataForPath](https://github.com/ncw/rclone/blob/b9e16b36e5208dc7460af59d9e75dc8234caca27/backend/onedrive/onedrive.go#L356) calls `f.dirCache.Get("")` and assumes the it got the right folder ID.

If `f.root` (`foldera` or `foldera/folderb` in the log below) does not exist, the returned folder ID is actually the drive root and `f.dirCache.foundRoot` is still false.

I assume the right function in this case should be `f.dirCache.FindRoot(false)` instead of `f.dirCache.Get("")`, but I'm not sure if this would break the issue in [onedrive.go#L379](https://github.com/ncw/rclone/blob/b9e16b36e5208dc7460af59d9e75dc8234caca27/backend/onedrive/onedrive.go#L379) again.

@Cnly any thoughts how this should be fixed?

#### What is your rclone version (output from `rclone version`)

`v1.47.0-032-gb9e16b36-beta`

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

`Linux Fabian-Desktop 5.0.9 #1-NixOS SMP Sat Apr 20 07:17:05 UTC 2019 x86_64 GNU/Linux`

####  Which cloud storage system are you using? (eg Google Drive)

OneDrive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone copy a.txt TestOneDrive:foldera`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
% rclone copy -vv a.txt TestOneDrive:
2019/05/03 23:44:20 DEBUG : rclone: Version "v1.47.0-032-gb9e16b36-beta" starting with parameters ["rclone" "copy" "-vv" "a.txt" "TestOneDrive:"]
2019/05/03 23:44:20 DEBUG : Using config file from "/home/fabian/.config/rclone/rclone.conf"
2019/05/03 23:44:21 DEBUG : a.txt: Couldn't find file - need to transfer
2019/05/03 23:44:21 DEBUG : a.txt: Starting multipart upload
2019/05/03 23:44:22 DEBUG : a.txt: Uploading segment 0/29 size 29
2019/05/03 23:44:22 INFO  : a.txt: Copied (new)
2019/05/03 23:44:22 INFO  :
Transferred:   	        29 / 29 Bytes, 100%, 15 Bytes/s, ETA 0s
Errors:                 0
Checks:                 0 / 0, -
Transferred:            1 / 1, 100%
Elapsed time:        1.8s

2019/05/03 23:44:22 DEBUG : 9 go routines active
2019/05/03 23:44:22 DEBUG : rclone: Version "v1.47.0-032-gb9e16b36-beta" finishing with parameters ["rclone" "copy" "-vv" "a.txt" "TestOneDrive:"]

% rclone copy -vv a.txt TestOneDrive:foldera
2019/05/03 23:44:35 DEBUG : rclone: Version "v1.47.0-032-gb9e16b36-beta" starting with parameters ["rclone" "copy" "-vv" "a.txt" "TestOneDrive:foldera"]
2019/05/03 23:44:35 DEBUG : Using config file from "/home/fabian/.config/rclone/rclone.conf"
2019/05/03 23:44:37 DEBUG : a.txt: Size and modification time the same (differ by -638.252406ms, within tolerance 1s)
2019/05/03 23:44:37 DEBUG : a.txt: Unchanged skipping
2019/05/03 23:44:37 INFO  :
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 1 / 1, 100%
Transferred:            0 / 0, -
Elapsed time:        1.3s

2019/05/03 23:44:37 DEBUG : 6 go routines active
2019/05/03 23:44:37 DEBUG : rclone: Version "v1.47.0-032-gb9e16b36-beta" finishing with parameters ["rclone" "copy" "-vv" "a.txt" "TestOneDrive:foldera"]

% rclone copy -vv a.txt TestOneDrive:foldera/folderb
2019/05/03 23:44:39 DEBUG : rclone: Version "v1.47.0-032-gb9e16b36-beta" starting with parameters ["rclone" "copy" "-vv" "a.txt" "TestOneDrive:foldera/folderb"]
2019/05/03 23:44:39 DEBUG : Using config file from "/home/fabian/.config/rclone/rclone.conf"
2019/05/03 23:44:41 DEBUG : a.txt: Size and modification time the same (differ by -638.252406ms, within tolerance 1s)
2019/05/03 23:44:41 DEBUG : a.txt: Unchanged skipping
2019/05/03 23:44:41 INFO  :
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 1 / 1, 100%
Transferred:            0 / 0, -
Elapsed time:        1.3s

2019/05/03 23:44:41 DEBUG : 6 go routines active
2019/05/03 23:44:41 DEBUG : rclone: Version "v1.47.0-032-gb9e16b36-beta" finishing with parameters ["rclone" "copy" "-vv" "a.txt" "TestOneDrive:foldera/folderb"]

% rclone tree TestOneDrive:foldera
2019/05/03 23:45:07 ERROR : : error listing: directory not found
2019/05/03 23:45:07 Failed to tree: directory not found
```

**Comment by ncw** [comment]
@Cnly are you going to have a go at fixing this?

**Comment by Cnly** [comment]
@ncw I haven't looked into it yet because I've been pretty busy... :( But I expect to have time for this in two or three days. Of course if you (or anybody else) have a fix already or have a plan to fix it, I'd really appreciate it!

**Comment by Cnly** [comment]
I'm able to reproduce this with the latest master version and OneDrive Personal, and thanks so much to @B4dM4n for digging into the problem. I think your analysis is correct that `readMetaDataForPath` shouldn't assume `dc.findRoot == true` even if `dc.Get("")` returns an ID. But I think instead of calling `dc.FindRoot(false)` in `readMetaDataForPath`, it may be better to call `dc.RootParentID()` and see if it returns an error. That way we don't need to have network traffic.

`dc.RootParentID()` returns an error immediately if any of the following is true:
- `!dc.findRoot`
- `dc.rootParentID == ""` (which is impossible because the code after [onedrive.go#379](https://github.com/ncw/rclone/blob/b9e16b36e5208dc7460af59d9e75dc8234caca27/backend/onedrive/onedrive.go#L379) only runs if root is inside a folder so it must have a parent)
- `dc.rootID == dc.trueRootID` (which is also impossible because root is inside a folder)

And if `dc.RootParentID()` returns an error, we know `dc.findRoot == false` and will proceed with the "query manually" branch.

Finally, there should also be a unit test for this.

@ncw @B4dM4n Would like to hear your comments.

**Comment by ncw** [comment]
The different betwen calling `dc.RootParentID()` and `dc.FindRoot(false)` is that the latter will actually find the root if needed.  The old code `dc.Get("")` didn't find the root so I think `dc.RootParentID()` is probably the better choice here.

**Comment by Cnly** [comment]
I've pushed a fix [here](https://github.com/ncw/rclone/commit/6e005b025a0fdcf7823ea0d4eb6a5c8cb585d4e9), but I'm unsure how to add the tests because it seems we need to access/modify contents under the absolute root of a backend.

**Comment by ncw** [comment]
> I've pushed a fix [here](https://github.com/ncw/rclone/commit/6e005b025a0fdcf7823ea0d4eb6a5c8cb585d4e9), but I'm unsure how to add the tests because it seems we need to access/modify contents under the absolute root of a backend.

Hmm, I try to keep the tests contained within a directory so you can use them on repositories with data you care about in.  Can we do a test without writing to the root?  If not then I think we'll have to skip it.

**Comment by Cnly** [comment]
I agree with you that we should not touch things outside the specified directory, and now I came up with a way to do the test.

Say we're testing using a remote whose and `remote.Root() == "user/given/path/rclone-test-1234567890"`. Then before the `FsMkdir` test finishes, we do

```go
deepRemote, err := fs.NewFs(remote.Name() + ":" + remote.Root() + "/deeper/nonexisting/directory")
require.NoError(t, err)

firstSlashIndex := strings.IndexRune(deepRemote.Root(), '/')
firstDir := deepRemote.Root()[:firstSlashIndex]
_, err = deepRemote.NewObject(firstDir)
require.Equal(t, fs.ErrorObjectNotFound, err)
```

The first two lines create a remote whose `dirCache.foundRoot == false`. The last four lines try getting the object at the absolute path `user/given/path/rclone-test-1234567890/deeper/nonexisting/directory/user`, which doesn't exist. And if the backend code is assuming `dirCache.Get("")` to be the ID of `user/given/path/rclone-test-1234567890/deeper/nonexisting/directory` (while it's actually the ID of the absolute root), `deepRemote.NewObject` will not return `fs.ErrorObjectNotFound` and the test will fail.

I tested with OneDrive before and after the fix and it worked. Now one thing I'm concerned about is whether there is any backend that returns `fs.ErrorObjectNotFound` in `NewObject` because it thinks "there _is_ a directory, not an object, so the _object_ is not found".
